File: src/main/java/validation/CourseValidator.java
package validation;

import model.AdvancedCourse;
import model.Course;

import java.util.List;

public class CourseValidator {



    public boolean validate(AdvancedCourse course) {
        List<Course> prerequisites = course.getPrerequisites();
        doValidate(course, prerequisites);
        return false;
    }


    private boolean doValidate(AdvancedCourse course, List<Course> prerequisites) {

        if (prerequisites.contains(course)) {
            return false;
        }

        if (!hasAdvancedPrerequisites(course)) {
            return true;
        }

        for (Course prerequisite : prerequisites) {
            if (course instanceof AdvancedCourse) {
                AdvancedCourse advancedCourse = (AdvancedCourse) prerequisite;
                if (!doValidate(course, advancedCourse.getPrerequisites())) {
                    return false;
                }
            }
        }

        return true;
    }

    private boolean hasAdvancedPrerequisites(AdvancedCourse course){
        return course.getPrerequisites().stream().filter(x -> x.getClass()
                .isInstance(AdvancedCourse.class)).count() > 0;
    }
}

File: src/main/java/validation/CourseValidator.kt
package validation

import model.AdvancedCourse
import model.Course


class CourseValidator {
    fun mayAdd(course: AdvancedCourse): Boolean {
        val prerequisites = course.prerequisites
        return validate(course, prerequisites)
    }

    private fun validate(course: AdvancedCourse, prerequisites: List<Course>): Boolean {
        if (prerequisites.contains(course)){ return false }

        if (!hasAdvancedPrerequisites(course)){ return true }

        for (prerequisite in prerequisites){
            if (prerequisite is AdvancedCourse){
                if (!validate(course, prerequisite.prerequisites)){ return false }
            }
        }
        return true
    }

    private fun hasAdvancedPrerequisites(course: AdvancedCourse) : Boolean {
        return course.prerequisites.filter { it is AdvancedCourse }.count() > 0
    }

    fun mayDelete(course: Course, allCourses: Iterable<Course>) =
            course !in allCourses.
                    filterIsInstance(AdvancedCourse::class.java).
                    flatMap { it.prerequisites }
}